### PR TITLE
Added exception catching for GeneratorsNeeded error in as_poly

### LINKED ISSUE #20610: Solve: GeneratorsNeeded with system involving constant equation
I don't think solve should convert all equations to Poly because it can be slow but at least there should be some checking for this case:
```python
>>> from sympy import *
>>> from sympy.abc import x, y
>>> solve([x + y, sqrt(2)], [x, y])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/enojb/current/sympy/sympy/sympy/solvers/solvers.py", line 1099, in solve
    solution = _solve_system(f, symbols, **flags)
  File "/Users/enojb/current/sympy/sympy/sympy/solvers/solvers.py", line 1774, in _solve_system
    subsol = _solve_system(subexpr, subsyms, **flags)
  File "/Users/enojb/current/sympy/sympy/sympy/solvers/solvers.py", line 1805, in _solve_system
    poly = g.as_poly(*symbols, extension=True)
  File "/Users/enojb/current/sympy/sympy/sympy/core/expr.py", line 1109, in as_poly
    poly = Poly(self, *gens, **args)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/polytools.py", line 162, in __new__
    return cls._from_expr(rep, opt)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/polytools.py", line 292, in _from_expr
    return cls._from_dict(rep, opt)
  File "/Users/enojb/current/sympy/sympy/sympy/polys/polytools.py", line 229, in _from_dict
    raise GeneratorsNeeded(
sympy.polys.polyerrors.GeneratorsNeeded: can't initialize from 'dict' without generators
```

### PR DESCRIPTION
**Pull request by kc611**
#### References to other Issues or PRs

Fixes #20610

#### Brief description of what is fixed or changed

In the `solve()` function, the piece of code which is supposed to return an empty list whenever a numerical constant is passed as an equation checks for `is_Number` in addition to `is_number`.

https://github.com/sympy/sympy/blob/dd9188277caada7161b64bfb56cf900b5480772b/sympy/solvers/solvers.py#L1002-L1024

But there are some instances like `sqrt(3)` which pass the `is_number` check (and according to the comment should return an empty list at this point) But the extra check of `is_Number` fails which leads to numerical constants like `sqrt(3)` or `I` to pass on as equations. Is there any significance of this `is_Number` check here or is it just an extra check to validate that argument is a number?

#### Release Notes

* polys
    * The `poly` function and `as_poly` methods now return `None` instead of raising `GeneratorsNeeded` for constant expressions when no generators given.
* solvers
    * Fixed bug causing `solve` to raise `GeneratorsNeeded` when equation such as `sqrt(2)` is included in a system.

### CODE REVIEW

**Comment by smichr** [comment]
I suspect that separating pure Numbers (like 1, 3.14, etc) from numerical expressions like (`cos(1)**2 + sin(1)**2 - 1`) was done because the former are easy to test for being zero or not. surds like `sqrt(3)` and number constants could also be included. But more complex "number" expressions cannot be classified as non-zero so trivially so if `is_zero` is not True we cannot conclude that there is no solution: we can only do so if `is_zero` is False (not None).

**Comment by kc611** [comment]
Ah, I see. I'll re-add the check once again.

I investigated a bit and it seems like this another issue altogether.

We can do :
```
>>>  Poly(sin(3),extension=True)
Poly((sin(3)), sin(3), domain='ZZ')
```
But doing this:
```
Poly(sqrt(3),extension=True)
```
Gives the error
```
sympy.polys.polyerrors.GeneratorsNeeded: can't initialize from 'dict' without generators
```

It seems that the culprit here is this piece of code here:

https://github.com/sympy/sympy/blob/dd9188277caada7161b64bfb56cf900b5480772b/sympy/polys/polyutils.py#L235-L237

`sin(3)` return this as `False` and `sqrt(3)` as `True` which leads this : 
  
https://github.com/sympy/sympy/blob/dd9188277caada7161b64bfb56cf900b5480772b/sympy/polys/polyutils.py#L257-L269

to believe that its a coefficient and thus does not register it as a generator

**Comment by oscarbenjamin** [comment]
If `sqrt(2)` is among the equations then there is no solution as easily verified by
```
In [202]: sqrt(2).is_zero
Out[202]: False
```
If we're checking `is_zero` then it should be easy to reject this case and so the output from solve should just be `[]`.

**Comment by kc611** [comment]
> I suspect that separating pure Numbers (like 1, 3.14, etc) from numerical expressions like (cos(1)**2 + sin(1)**2 - 1) was done because the former are easy to test for being zero or not. 

@smichr I just found out that when we do add complex numerical expressions like (cos(1)**2 + sin(1)**2 - 1) `solve` is unable to solve it out as a expression reducible to zero : 

```
from sympy import *

x = Symbol('x')
y = Symbol('y')

print(solve([x-y, 0],[x,y]))
>>> {x:y}
print(solve([x-y, sin(2)**2 + cos(2)**2],[x,y]))
>>> []
print(solve([x-y, sin(2)**2 + cos(2)**2 - 1],[x,y]))
>>> [] # Wrong
```
Also this leads to the numerical expression being passed down and converted into `Poly` which should not be allowed in the first place (Maybe this https://github.com/sympy/sympy/pull/20661#issuecomment-751279211 should be filed as another issue). 

As of now I think we should add a Numerical evaluation over here something like `.evalf(chop=True)`  in addition to `is_zero()`. And replace that numerical expression with zero if this checks as true to prevent numerical constants being converted into instances of `Poly`

**Comment by oscarbenjamin** [comment]
We need to distinguish the three cases for `is_zero`: True, False and None.

In the True case the equation can be ignored. In the False case the whole system has no solutions. In the None case we don't know whether there are solutions or not. At that point we either need to treat it as being True or as being False or we need to try harder to find out whether it is zero or not e.g. by simplifying.

It isn't really necessary to try numerical evaluation because that already happens implicitly in `is_zero` at least in the real case.

**Comment by kc611** [comment]
> We need to distinguish the three cases for is_zero: True, False and None.

I presume by this you mean the cases where 
i. the numerical evaluation is sufficiently large (so obviously non-zero)  classified as `False` hence no solution, 
ii. numerical evaluation is outright zero ( which will be caught by `.is_zero` check ) classified as `True`  hence ignore equation 
iii. numerical evaluation is extremely small but not zero (for example in case of N(cos(1)**2 + sin(1)**2 - 1) ) will be `None`.

I don't know of any function in sympy which could simplify a expression that passes the `.is_number` check other than simple numerical evaluation. 

I think we could set a threshold for `None` cases, so that if the numerical evaluation is above this is threshold classify it as non-zero and zero otherwise.

**Comment by oscarbenjamin** [comment]
> I think we could set a threshold for `None` cases, so that if the numerical evaluation is above this is threshold classify it as non-zero (hence `False`) and zero (hence `True`) otherwise.

This is already what happens internally when you check `is_zero`:
```python
In [9]: class A(Expr):
   ...:     def _eval_evalf(self, prec):
   ...:         return Float(1, prec)
   ...: 

In [10]: a = A()

In [11]: a.is_zero
Out[11]: False
```

> I don't know of any function in sympy which could simplify a expression that passes the `.is_number` check other than simple numerical evaluation.

Numerical evaluation can show when an expression is nonzero but it is not usually possible to prove that an expression is zero (rather than small) using numerical evaluation. The obvious function that can resolve a case like this is `simplify` but that's potentially very expensive:
```
In [13]: e = (cos(1)**2 + sin(1)**2) - 1

In [14]: e
Out[14]: 
        2         2   
-1 + cos (1) + sin (1)

In [15]: print(e.is_zero)
None

In [16]: simplify(e)
Out[16]: 0
```
We generally want to avoid using `simplify` but that means that some cases go unresolved.

File: sympy/solvers/solvers.py [hunk 1]
```diff
@@ -1018,11 +1018,9 @@ def _sympified_list(w):
             ok = True
         else:
             if fi.is_number:
-                if fi.is_Number:
-                    if fi.is_zero:
-                        continue
+                fi = simplify(fi)
+                if not fi.is_zero:
```
Comment (kc611) [inline sympy/solvers/solvers.py:1022]: This is what I propose. Here even if `simplify` is computationally expensive, doing it this way will save us computation by handling all the numerical cases in this block of code and not allowing them to propogate further (and convert into `Poly`) .

Since when expression evaluates to `zero` it'll have no effect on the system as a whole so it can be omitted. If it is non-zero even after a `simplify` call we'll simply return it as an empty list.

Comment (oscarbenjamin) [reply sympy/solvers/solvers.py:1022]: Sometimes `simplify` can be extremely slow and might even hang. It is possibly better just to give the benefit of the doubt if we are not sure that the expression is zero or not and assume that it is.

Comment (kc611) [reply sympy/solvers/solvers.py:1021]: Alright I'll remove that but on its own `.is_zero` is not able to handle this not sure condition. ( As in the case of sin(2)**2 + cos(2)**2 - 1 which I presume internally `is_zero` evaluates numerically and is unable to classify it as zero ).

Should I instead do a `.evalf(chop = True)` or let this be.

Comment (oscarbenjamin) [reply sympy/solvers/solvers.py:1021]: I think just let it be. If the equation is not known to be zero or not we can allow it to progress to the next stages of solving.

**Comment by kc611** [comment]
@oscarbenjamin Any suggestions? This PR looks ready.

File: sympy/solvers/solvers.py [hunk 2]
```diff
@@ -1018,11 +1018,8 @@ def _sympified_list(w):
             ok = True
         else:
             if fi.is_number:
-                if fi.is_Number:
-                    if fi.is_zero:
-                        continue
+                if not fi.is_zero:
```
Comment (oscarbenjamin) [inline sympy/solvers/solvers.py:1021]: Can this be `if fi.is_zero is False`?

Comment (kc611) [reply sympy/solvers/solvers.py]: Alright.

**Comment by smichr** [comment]
I think you are no longer ignoring equations which are identically 0. By not setting `ok = True` then aren't non-zero-is_number = None expressions ignored?

**Comment by kc611** [comment]
> I think you are no longer ignoring equations which are identically 0. 

I do not understand. The way current logic is written, if `is_zero` evaluates to `True` then the equation is ignored since the `ok` variable will retain it's `False` value from [here](https://github.com/sympy/sympy/blob/b0b16e164d35802271bc6c46c79c7ef174ea0c3a/sympy/solvers/solvers.py#L1016) which will then fail [this](https://github.com/sympy/sympy/blob/b0b16e164d35802271bc6c46c79c7ef174ea0c3a/sympy/solvers/solvers.py#L1029) check and hence not be appended in the `newf` list. (Hence ignored)

> By not setting ok = True then aren't non-zero-is_number = None expressions ignored?

The `ok = True` check I removed was initially meant for numerical expressions which do pass the `is_number` check but evaluate to zero.(To prevent them to be converted into Poly class, since their presence won't affect the solution of the system as a whole) I shouldn't affect the equations which give `is_number` as `None` since those equations weren't entering that part of code in the first place.

**Comment by oscarbenjamin** [comment]
I think the point is that when `is_zero` gives `None` we should allow the equation to progress to the next stage of solving where further manipulations might discover that it is or is not zero.

**Comment by kc611** [comment]
I see. But as pointed out here https://github.com/sympy/sympy/pull/20661#issuecomment-751373735 it seems that further simplification simply treats those equations as non-zero ( or rather it shows a behaviour quite similar to `is_zero` check ). IMO  classifying those corner cases as zero as we have now and solving rest of the system is well worth the against the letting them pass into `Poly` and risk such issues happening again (Unless of course, those aren't corner cases and there are considerable amount of equations being wrongly classified to zero which the further simplification actually prevents, In that case it would be better to let that check be.).

As of now I could not find any equation which is actually zero and gives `None` to the `is_zero` check,yet isn't treated as a non-zero equation which seems to be the default behavior of `solve`

**Comment by oscarbenjamin** [comment]
The exception from `Poly` could just be caught if that is the problem.

**Comment by kc611** [comment]
I re-added the `ok = True` check

File: sympy/solvers/solvers.py [hunk 3]
```diff
@@ -1018,11 +1018,10 @@ def _sympified_list(w):
...
         else:
             if fi.is_number:
-                if fi.is_Number:
-                    if fi.is_zero:
-                        continue
+                if fi.is_zero is False:
                     return []
-                ok = True
+                if not fi.is_Number is True:
+                    ok = True
```
Comment (oscarbenjamin) [inline sympy/solvers/solvers.py:1024]: I find this change quite hard to parse. The reason for using `is False` with `fi.is_zero` is because it has three possible values (True, False and None). That isn't the case for `is_Number` which is always True or False.

Otherwise I'm not sure I follow what the changed logic here is doing. Previously this would `continue` if `fi.is_zero` is true. It doesn't look like it still does that.

Comment (kc611) [reply sympy/solvers/solvers.py:1024]: > Previously this would `continue` if fi.is_zero is true. It doesn't look like it still does that.

I think the keyword you're mentioning is `pass`. Instead `continue` keyword will skip the rest of the code inside a loop for the current iteration. 

https://github.com/sympy/sympy/blob/17cf0f5ad8c2535e32009869220654e7fd25865f/sympy/solvers/solvers.py#L1016-L1034

Which means previously if the expression `is_Number` and `is_zero` then the loop will continue onto it's next iteration without the `ok = True` (#L1025) or the `newf.append(fi)` (#L1030) statements being executed for the current iteration. (This might be because `is_zero` would be either `True` or `False` for are `Number` instances). This is as same as the current logic where the expression is being ignored if it `is_Number` and `is_zero`.

Comment (oscarbenjamin) [reply sympy/solvers/solvers.py:1024]: Can we just rewrite the whole loop as:
```python
if fuzzy_or(fuzzy_not(fi.is_zero) for fi in f):
    return []

def ok(fi):
    if fi.free_symbols & symset:
        return True
    else:
        return fi.is_number or fi.is_constant()

f = [fi for fi in f if ok(fi)]
```
I guess that the intention is to try and avoid checking `is_zero` unless absolutely necessary but it could be simpler like:
```python
for fi in f:
    if fi.free_symbols & symset:
        newf.append(fi)
    elif fi.is_number:
        if fi.is_zero is False:
            return []
        else:
            newf.append(fi)
    elif fi.is_constant():
        newf.append(fi)
    else:
        # Skip this equation
        continue
```

Comment (kc611) [reply sympy/solvers/solvers.py]: I tried implementing your suggested loop but it still seems to be failing tests. It seems to me that there are instances of `Number` such as `Matrix` (one of the failing tests ) which are not at all meant to be passed in `Poly` class. Hence initially (even before this PR) the logic was such that it handled at least all of the `Number` instances in this loop itself rather than letting them pass. This seems to be different for the expressions which are `number` but not `Number`, and they can be given forward for further processing. The current issue for this PR seems to have arisen from the same whence a `sqrt` which was supposed to be a `Number` but isn't was passed on forward into `Poly` class.

Comment (oscarbenjamin) [reply sympy/solvers/solvers.py]: I don't think that we should use the loop here as a way of preventing an expression from being passed to `Poly` later. This loop here is to try and handle equations not involving the unknowns based on whether or not the are known to be zero and that is the only role it should play.

The problem with `Poly(sqrt(2))` raising should be handled at the place where `Poly` is called.

Comment (kc611) [reply sympy/solvers/solvers.py]: You're right there are too many variables here we can just catch this error in as_poly(), similar to how PolynomialError is being caught there. 

This'll also handle the cases when user tries to create objects like `Poly(sqrt(2))` seperately.

**Comment by kc611** [comment]
@oscarbenjamin I added the exception catch as requested.

File: sympy/core/expr.py [hunk 4]
```diff
@@ -1116,6 +1116,8 @@ def as_poly(self, *gens, **args):
                 return poly
         except PolynomialError:
             return None
+        except GeneratorsNeeded:
+            return None
```
Comment (oscarbenjamin) [inline sympy/core/expr.py:1120]: You can combine this:
```
except (PolynomialError, GeneratorsNeeded):
```
It would be good to add a comment here explaining what would cause either of these exceptions to be raised and why they are being caught.

If this is changing `as_poly` then there should be tests added for `as_poly` in `test_expr.py`.

Comment (kc611) [reply sympy/core/expr.py]: @oscarbenjamin I added the changes as requested.

File: sympy/core/expr.py [hunk 5]
```diff
@@ -1114,7 +1114,10 @@ def as_poly(self, *gens, **args):
                 return None
             else:
                 return poly
-        except PolynomialError:
+        except (PolynomialError, GeneratorsNeeded):
+            # GeneratorsNeeded will be raised here when surds when constant non-algebraic surds are being passed
+            # into as_poly and is being caught here to ensure smooth operation of solve in case it raises.
+            # See Issue #20610
```
Comment (oscarbenjamin) [inline sympy/core/expr.py:1120]:
```suggestion
        except (PolynomialError, GeneratorsNeeded):
            # PolynomialError is caught for e.g. exp(x).as_poly(x)
            # GeneratorsNeeded is caught for e.g. S(2).as_poly()
```

File: sympy/core/tests/test_expr.py [hunk 6]
```diff
@@ -1130,6 +1130,8 @@ def test_as_poly_as_expr():
     p = Poly(f, x, y)
 
     assert p.as_poly() == p
+    # Issue #20610
+    assert (sqrt(2).as_poly(extension = True)) == None
```
Comment (oscarbenjamin) [inline sympy/core/tests/test_expr.py:1134]:
```suggestion
    # https://github.com/sympy/sympy/issues/20610
    assert S(2).as_poly() is None
    assert sqrt(2).as_poly(extension=True) is None
```

**Comment by kc611** [comment]
@oscarbenjamin All the suggested changes have been applied.

**Comment by kc611** [comment]
@oscarbenjamin The suggested changes have been applied. Have a look

**Comment by oscarbenjamin** [comment]
The changes look good.

Can you squash the commits and give a single commit message summarising what is changed e.g.:
```
fix(polys): make poly and as_poly return None for constants without generators
```

The release note could be:
```
* polys
    * The `poly` function and `as_poly` methods now return `None` instead of raising `GeneratorsNeeded` for constant expressions when no generators are given.
 * solvers
    * Fixed a bug causing `solve` to raise `GeneratorsNeeded` when an equation such as `sqrt(2)` is included in a system.
```

**Comment by oscarbenjamin** [comment]
Something definitely went wrong when squashing the commits. The PR now shows 153 files changed.

**Comment by kc611** [comment]
I think I mistakenly reverted to a previous commit. I'll delete this branch and open a new PR. Sorry for the inconvenience.
